extensions/ql-vscode/src/codeql-task-provider.ts

```diff
@@ -0,0 +1,97 @@
+import * as vscode from 'vscode';
+import { CodeQLCliServer } from './cli';
+
+interface CodeQLTaskDefinition extends vscode.TaskDefinition {
+  /**
+   * The command to run.
+   */
+  command: string;
+
+  /**
+   * Additional command arguments.
+   */
+  commandArgs?: string[];
+}
+
+export class CodeQLTaskProvider implements vscode.TaskProvider {
+  // Matches the label in package.json
+  static CodeQLType = 'codeql';
+  private tasks: vscode.Task[] | undefined;
+  private sharedState: string | undefined;
+
+  constructor(private workspaceRoot: string, private cliServer: CodeQLCliServer) { }
+
+  public async provideTasks(): Promise<vscode.Task[]> {
+    return this.getTasks();
+  }
+
+  public resolveTask(_task: vscode.Task): vscode.Task | undefined {
+    const command: string = _task.definition.command;
+    if (command) {
+      const definition: CodeQLTaskDefinition = <any>_task.definition;
+      return this.getTask(definition.command, definition.commandArgs ? definition.commandArgs : [], definition);
+    }
+    return undefined;
+  }
+
+  private getTasks(): vscode.Task[] {
+    if (this.tasks !== undefined) {
+      return this.tasks;
+    }
+    const commands: string[] = ['resolve languages', 'version'];
+    // Each command can have some commandArgs.
+    const commandArgs: string[][] = [[]];
+
+    this.tasks = [];
+    commands.forEach(command => {
+      commandArgs.forEach(commandArgsGroup => {
+        this.tasks!.push(this.getTask(command, commandArgsGroup));
+      });
+    });
+    return this.tasks;
+  }
+
+  private getTask(command: string, commandArgs: string[], definition?: CodeQLTaskDefinition): vscode.Task {
+    if (definition === undefined) {
+      definition = {
+        type: CodeQLTaskProvider.CodeQLType,
+        command,
+        commandArgs
+      };
+    }
+    return new vscode.Task(definition, vscode.TaskScope.Workspace, `${command} ${commandArgs.join(' ')}`,
+      CodeQLTaskProvider.CodeQLType, new vscode.CustomExecution(async (): Promise<vscode.Pseudoterminal> => {
+        // When the task is executed, this callback will run. Here, we setup for running the task.
+        return new CodeQLTaskTerminal(this.workspaceRoot, this.cliServer, command, commandArgs, () => this.sharedState, (state: string) => this.sharedState = state);
+      }));
+  }
+}
+
+class CodeQLTaskTerminal implements vscode.Pseudoterminal {
```

Comment:
I guess there's no way to just use the regular extension logger for output and avoid creating a pseudo terminal? I feel like that would be nicer, but not sure if that's possible.

Reply:
I think this should be possible 🤔  Will do some experimenting!

Reply:
I couldn't work out how to use the regular logger, so I've worked on fixing the pseudo terminal so that it now actually runs the task 😅